xpi: add rsa key cache

jvehent opened the pull request:
Signing XPIs is currently slow due to the RSA key generation required for every signature. This patch adds a cache of 100 private keys that gets populated in the background to accelerate signing requests.

jvehent commented:
Consuming from a channel is thread safe. For example, those 5 goroutines are consuming from the same channel and the same value never gets read twice. https://play.golang.org/p/jxCioPPvT_r

autrilla commented on signer/xpi/x509.go:
```diff
@@ -7,12 +7,63 @@ import (
+			log.Fatalf("xpi.populateRsa2048Cache: %v", err)
+		}
+		s.rsa2048Cache <- key
+		time.Sleep(time.Minute)
```
The channel will block when writing if it's full, so I don't think this is needed

jvehent replied:
I we remove this, we'll deplete the entropy when autograph starts, which could slow down other types of signing. This is a way to play nice with others.

autrilla replied:
Oh, nice! makes sense

autrilla commented on signer/xpi/x509.go:
```diff
@@ -7,12 +7,63 @@ import (
+		select {
+		case key := <-s.rsa2048Cache:
+			return key, nil
+		case <-time.After(100 * time.Millisecond):
```
If the sleep when generating keys is removed, I don't think this is needed

autrilla commented on signer/xpi/x509.go:
```diff
@@ -7,12 +7,63 @@ import (
+			// generate a key if none available
+			return rsa.GenerateKey(rand.Reader, size)
+		}
+	case 4096:
```
nit: perhaps cleaner to keep a `map[int]chan *rsa.PrivateKey` to look up the caches instead of duplicated code

probably overkill for two key sizes, but maybe more sizes will be added in the future?

jvehent replied:
agreed, but indeed something I'd do in a refactoring if we end up needed more key sizes. I think we'll move to ecdsa before that though.

jvehent commented:
I think we can trust clients because we issue hawk tokens to them. The
service is also internal only, and not opened to the internet.

On Thu, Apr 5, 2018, 20:56 J.C. Jones <notifications@github.com> wrote:

> *@jcjones* commented on this pull request.
>
> Can this ever be used by untrusted clients, or can we assume clients are
> trustworthy? Trying to make sure my threat model here is correct.
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> <https://github.com/mozilla-services/autograph/pull/74#pullrequestreview-109921706>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAZXgXKUK19aTLMblixV3erSRDOn5QjTks5tlr1DgaJpZM4TI6Gn>
> .
>

jvehent commented:
Alright, thanks everyone for the reviews! I refactored it one last time because I realized the cache knows the key size at init, since it matches the issuer private key. I also added unit test, and discovered `t.Parallel()`, which is pretty awesome.

@g-k: could I get one last pass when you get a chance please?
